### DISCUSSION TITLE
[Question] What difference between `Async` and `Sync`

### DISCUSSION THREAD

**Discussion by Mark24Code** [socketry/async#285] [Q&A] [answered]
I wish the document with more examples and details explanation.  😃

**Comment by trevorturk** [comment]
Likely this should be moved to a Discussion, but a couple quick links that may help explain things: 

* https://socketry.github.io/async/guides/best-practices/index#use-a-top-level-sync-to-denote-the-root-of-your-program
* https://github.com/socketry/async/issues/237

I bet additional documentation would be welcome, if you think that's something you could help with after reading through these docs!

**Comment by ioquatix** [comment]
- `Async{...}` returns a promise and executes your code in an event loop.
- `Sync{...}` returns a value of executing your code in an event loop.

Does this clarify it? We can certainly look at improving the documentation.

**Reply by hibachrach** [reply]
I think it might help to see how it compares with `async`/`await` in other languages. Does `Kernel#Sync` have a correspondence in JavaScript? From your description it sounds like an `await` of an immediately invoked promise.

**Reply by ioquatix** [reply]
Yes, that's correct, e.g. the following Ruby:

```ruby
Sync{bar()} # returns value of bar()
```

Is similar to the following JavaScript:

```javascript
await bar(); // returns value of bar()
```

And for completeness:

```ruby
Async{bar()} # returns a promise (task)
```

Is similar to:

```javascript
bar(); // returns a promise
```

There are some minor semantic differences that make `Sync` a bit more important in Ruby, notably, that without `Sync` there is no event loop, while in JavaScript the event loop always exists (from process inception). Hope that helps.

**Reply by hibachrach** [reply]
Honestly, I think a "JS -> Ruby" translation example in the docs would do a lot. JS's `async`/`await` (and similar syntaxes in Python, etc.) are fairly ubiquitous. But I also wonder if `Sync` would be better called `AwaitAsync` or even remove that as a convenience method and just have only `Async` and forcing you to call `await` on it, like how Rust does it.

**Reply by ioquatix** [reply, marked as answer]
`Sync` was a joke that became useful, unfortunately hard to change the name now :)

I think improving the documentation is a good idea: https://github.com/socketry/async/pull/465

**Reply by hibachrach** [reply]
TYSM! The new docs look great. 

> a joke that became useful

I think i'll use this to describe myself from now on
